### PR TITLE
Add Hypertest scenarios

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Adds functional tests of the Hydra API.

### CODE REVIEW

File: .github/hypertest.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+function finish() {
+  make logs stop
+}
+
+trap finish EXIT
+
+make start wait-healthy
+
+docker run --rm --network host --mount "type=bind,source=$(pwd)/test/hypertest/,destination=/tests" hydrofoil/hypertest:_0.3.2 --baseUri http://localhost:8080/
```
Comment (thewilkybarkid) [inline .github/hypertest.sh]: Not currently running anything, also seeing:

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$2 (file:/hydra-validator/node_modules/@hydrofoil/hypertest/jdeploy-bundle/app.hypermedia.testing.dsl.cli-1.0.0-SNAPSHOT-shaded.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$2
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

/cc @tpluscode

Comment (tpluscode) [reply .github/hypertest.sh]: Yes, I'm aware of that. Haven't investigated yet but it's something coming from the internals of the DSL code.

Comment (thewilkybarkid) [reply .github/hypertest.sh]: Ok, so the warning is unrelated to the scenario not being found?

Comment (tpluscode) [reply .github/hypertest.sh]: Yes. I think the version 0.3.2 did not like to have the tests mounted.

Please try `latest` where I switched to recommending `volumes` as suggested by my colleague. I will tag that as 0.4 soon

```
version: "3"

services:
  e2e-tests:
    image: "hydrofoil/hypertest"
    network_mode: "host"
    environment:
      BASE_URI: "http://localhost:8080/"
    volumes:
      - ./test/hypertest:/tests
```

File: .gitignore [hunk 2]
```diff
@@ -1,2 +1,3 @@
 /build/
 /node_modules/
+/test/hypertest/*.hypertest.json
```
Comment (thewilkybarkid) [inline .gitignore]: Would rather the compiled files weren't in the same directory.

**Comment by thewilkybarkid** [comment]
Currently seeing

```
Failed to dereference http://localhost:8080/ FetchError: request to http://localhost:8080/ failed, reason: socket hang up
```

in CI, and locally it just hangs (and can't SIGTERM out).

**Comment by thewilkybarkid** [comment]
Works outside of Docker.

File: .gitignore (hunk 2 above)
Comment (thewilkybarkid) [reply .gitignore]: https://github.com/hypermedia-app/hypertest-docker/issues/1

**Comment by thewilkybarkid** [comment]
@giorgiosironi I'm seeing different behaviour on MacOS to GitHub Actions (neither actually working though). What do you see?

**Comment by giorgiosironi** [comment]
My output: failure, but the networking seems to work.

```
------
   Compiling test scenarios
   Directory used: /tests
   Using temp directory for compiled output: /tmp/d-20191111-13-3523ti.z3n1i
------

WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$2 (file:/hydra-validator/node_modules/@hydrofoil/hypertest/jdeploy-bundle/app.hypermedia.testing.dsl.cli-1.0.0-SNAPSHOT-shaded.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$2
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Compiling article-list.hydra
Done!

------
   Running scenario article-list
------

2019-12-11T09:42:51.143Z INFO      Analysis started...
2019-12-11T09:42:51.221Z INFO      Fetched resource http://localhost:8080/
2019-12-11T09:42:51.222Z SUCCESS   Found expected property http://schema.org/name
2019-12-11T09:42:51.223Z WARNING   Stepping into link http://www.w3.org/ns/hydra/core#collection Resources found but http://www.w3.org/ns/hydra/core#collection is not a SupportedProperty of hydra:Link type.
2019-12-11T09:42:51.263Z INFO      -- Fetched resource http://localhost:8080/articles
2019-12-11T09:42:51.264Z SUCCESS   -- Found expected property http://www.w3.org/ns/hydra/core#title
2019-12-11T09:42:51.264Z INFO      -- Stepping into property http://www.w3.org/ns/hydra/core#manages
2019-12-11T09:42:51.264Z SUCCESS   ---- Found expected property http://www.w3.org/ns/hydra/core#property
2019-12-11T09:42:51.264Z SUCCESS   ---- Found expected property http://www.w3.org/ns/hydra/core#object
2019-12-11T09:42:51.264Z SUCCESS   -- Found expected property http://www.w3.org/ns/hydra/core#totalItems
2019-12-11T09:42:51.264Z INFO      -- Stepping into members of array http://www.w3.org/ns/hydra/core#member
2019-12-11T09:42:51.264Z FAILURE   Executed 10 out of 12 steps. Strict mode requires that all steps are executed
2019-12-11T09:42:51.265Z FAILURE   Analysis complete with errors Success: 5, Warnings: 1, Failures: 1

------
   Summary
------

0/1 scenarios succeeded.

------
   Failed scenarios
------

  - article-list
```

**Comment by giorgiosironi** [comment]
I suspected some issue with the Docker containers accessing the host network. Introducing a smoke test for that changed the result, so there may be some race conditions in the bootstrap.

**Comment by thewilkybarkid** [comment]
Seems to be working now (ie Hypertest is running; both locally and on Actions). Not sure why.

**Comment by giorgiosironi** [comment]
Hypothesis is the smoke test stalls the process long enough for everything to be wired up and `localhost:8080` to lead to the right place. But still the checks are failing?

**Comment by thewilkybarkid** [comment]
Seems fine without.

**Comment by thewilkybarkid** [comment]
Warning is https://github.com/libero/article-store/pull/49.

**Comment by thewilkybarkid** [comment]
Main Hypertest failure relates to https://github.com/libero/article-store/blob/a0d83b491711984713c3afd2406b78f2c1e815d2/test/hypertest/article-list.hydra#L16-L19 but there are none and strict mode is enabled. Adding the operation should fix it.

**Comment by giorgiosironi** [comment]
> Seems fine without.

Could have been a unlucky original run, not sure if it reproduced easily.

**Comment by giorgiosironi** [comment]
I'll disappear from here until ready for review :+1:

**Comment by thewilkybarkid** [comment]
Things I've learnt:

- `204 No Content` isn't suitable following a `POST` as there's no links to follow (this was only temporary anyway, as current the articles aren't separate resources)
- Named operations are good. I've introduced Schema.org action types, but proper identified actions is probably better (eg `http://libero.pub/article-store/add-article`).
- It doesn't look feasible to do tests like 'add an article then it should appear in the article list', but think that's ok (and handled by the other types of tests).
- Probably don't want to check most properties are present (currently check for the existence of 'hydra:property' in 'hydra:manages' on a 'hydra:Collection'), instead rely on the other tests.

**Comment by thewilkybarkid** [comment]
Still got one warning related to the add article operation: `Could not determine the resource representation`. Not sure what it's referring too.

Also see `No processor found for media type text/plain; charset=utf-8` in the logs.

**Comment by tpluscode** [comment]
> Could not determine the resource representation

 It means that the resource graph did not contain the resource identifier by the dereferenced URI.

For example, you request `http://example.app/resource-one?q=foo` but got

```
<http://example.app/resource-two> schema:name "I'm not resource one" ;
<http://example.app/resource-one> schema:name "I am similar but not exactly the same"
```

The client [alcaeus](https://alcaeus.hydra.how) looks for the `/resource-one?q=foo` resource with exact string comparison. It will not find the one without query params, hence the warning.

Please refer to [Root resource](https://alcaeus.hydra.how/latest/representations/root.html) page in alcaeus' docs for some ways it tries to figure out the right resource to return from the graph

I might have rephrase this message. Ideas welcome.

**Comment by thewilkybarkid** [comment]
> It means that the resource graph did not contain the resource identifier by the dereferenced URI.

Seems to be the `201 Created` response to the `POST`, which doesn't contain a body. (The docs do define `hydra:returns` as `owl:nothing`.)

**Comment by tpluscode** [comment]
The docs notwithstanding, I think I could skip this warning if the body is indeed empty. Or at least in empty 204 and 201 responses.

**Comment by thewilkybarkid** [comment]
> Also see `No processor found for media type text/plain; charset=utf-8` in the logs.

Just noticed the Generic Hydra Application emits the same following a `POST` with no body.

Edit: Just seen that Koa is returning 'Created' as the body, rather than no body at all...

File: src/routes/api-documentation.ts [hunk 3]
```diff
@@ -21,7 +21,7 @@ export default (): AppMiddleware => (
           '@type': hydra.Class,
           [hydra.title]: { '@value': 'API entry point', '@language': 'en' },
           [hydra.supportedOperation]: {
-            '@type': hydra.Operation,
+            '@type': [hydra.Operation, schema.ViewAction],
```
Comment (thewilkybarkid) [inline src/routes/api-documentation.ts]: Not sure if `ViewAction` is the most appropriate. `DownloadAction`? `ReceiveAction`?

**Comment by thewilkybarkid** [comment]
Will mark as ready when a new version of `hydrofoil/hypertest` is tagged.

**Comment by thewilkybarkid** [comment]
Might as well review now, but won't merged until a released is tagged. /cc @tpluscode

**Comment by tpluscode** [comment]
Anything still missing/not working for you? If not they let me tag the current state

**Comment by thewilkybarkid** [comment]
Works fine, we can carry on using `--init`  to get round https://github.com/hypermedia-app/hypertest-docker/issues/2.

**Comment by tpluscode** [comment]
Tagged as 0.4.0. Docker hub is going to build it soon

File: src/routes/api-documentation.ts (hunk 3 above)
Comment (thewilkybarkid) [reply src/routes/api-documentation.ts]: Moving to https://github.com/libero/article-store/pull/128.

**Pull request merged**